src/cancelq/posix/cancellable.c

```diff
@@ -0,0 +1,399 @@
+    }
+    switch (q->ids[id])
+    {
+    case ID_UNUSED:
+        return FAIL_NO_SUCH_ID;
+    case ID_WAITING:
+        return FAIL_NOT_RUN;
+    case ID_DELIVERED:
+        break;
+    }
+    q->ids[id] = ID_UNUSED;
+    return SUCCESS;
+}
+
+static cancellation_status_t cancel_LH(event_queue_t *q, cancellable_id_t id,
+                                       void **event)
+{
+    size_t idx;
+
+    if (q->idsize < id)
+    {
+        return FAIL_NO_SUCH_ID;
+    }
+
+    switch (q->ids[id])
+    {
+    case ID_UNUSED:
+        return FAIL_NO_SUCH_ID;
+    case ID_WAITING:
+        break;
+    case ID_DELIVERED:
+        return FAIL_ALREADY_RUN;
+    }
+
+    // This is pretty deep into the heap's data structure, but heaps
+    // don't do this.
+    for (idx = 1; idx < q->heap.nextidx; idx++)
+    {
+        if (q->heap.data[idx].id == id)
+        {
+            *event = (void *)q->heap.data[idx].event.event;
+            heap_rm(&q->heap, idx);
+            q->ids[id] = ID_UNUSED;
+            return SUCCESS;
+        }
+    }
+    assert(false); // This should never ever happen; it's waiting but corrupted.
+}
+
+cancellation_status_t eq_cancel(event_queue_t *q, cancellable_id_t id,
+                                void **event)
+{
+    cancellation_status_t status;
+
+    status = FAIL_LOCKING;
+    *event = NULL;
+    if (pthread_mutex_lock(&q->lock) == 0)
+    {
+        status = cancel_LH(q, id, event);
+        pthread_mutex_unlock(&q->lock);
+    }
+
+    return status;
+}
+
+cancellation_status_t eq_cancel_or_release(event_queue_t *q,
+                                           cancellable_id_t id,
+                                           void **event)
+{
+    cancellation_status_t status;
+
+    status = FAIL_LOCKING;
+    *event = NULL;
+    if (pthread_mutex_lock(&q->lock) != 0)
+        goto end;
+    if (q->idsize < id)
+    {
+        status = FAIL_NO_SUCH_ID;
+        goto unlock;
+    }
+
+    switch (q->ids[id])
+    {
+    case ID_UNUSED:
+        status = FAIL_NO_SUCH_ID;
+        break;
+    case ID_WAITING:
+        status = cancel_LH(q, id, event);
+        break;
+    case ID_DELIVERED:
+        status = release_LH(q, id);
+        break;
+    }
+
+unlock:
+    pthread_mutex_unlock(&q->lock);
+end:
+    return status;
+}
+
+bool eq_validate(event_queue_t *q)
+{
+    bool status;
+
+    status = false;
+    if (pthread_mutex_lock(&q->lock) == 0)
+    {
+        status = check_queue(q);
+        pthread_mutex_unlock(&q->lock);
+    }
+    return status;
+}
+
+cancellation_status_t eq_release(event_queue_t *q, cancellable_id_t id)
+{
+    cancellation_status_t status;
+
+    status = FAIL_LOCKING;
+    if (pthread_mutex_lock(&q->lock) == 0)
+    {
+        status = release_LH(q, id);
+        pthread_mutex_unlock(&q->lock);
+    }
+
+    return status;
+}
+
+void eq_wait_empty(event_queue_t *q)
+{
+    pthread_mutex_lock(&q->lock);
+    while(!empty_LH(q))
+    {
+        pthread_cond_wait(&q->empty, &q->lock);
+    }
```

Comment:
Won't this deadlock pretty much always?  I think you may need to unlock and relock around the `pthread_cond_wait`.

Reply:
From the man page:

       pthread_cond_wait    atomically    unlocks    the    mutex    (as   per
       pthread_unlock_mutex) and waits for the condition variable cond  to  be
       signaled.  The  thread  execution is suspended and does not consume any
       CPU time until the condition variable is signaled. The  mutex  must  be
       locked  by  the calling thread on entrance to pthread_cond_wait. Before
       returning to the calling thread,  pthread_cond_wait  re-acquires  mutex
       (as per pthread_lock_mutex).